File: Heaters3.py
class Solution:
    def findRadius(self, houses, heaters) -> int:
        houses.sort()
        heaters.sort()
        result = 0
        index = 0
        for house in houses:
            left = index
            right = len(heaters) - 1
            while left <= right:
                mid = (left + right) // 2
                if heaters[mid] < house:
                    left = mid + 1
                else:
                    right = mid - 1
            index = left
            dist1 = float('inf') if left == len(heaters) else heaters[left] - house
            dist2 = float('inf') if left == 0 else house - heaters[left - 1]
            result = max(result, min(dist1, dist2))
        return result


if __name__ == '__main__':
    houses = [int(num) for num in input().split()]
    heaters = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.findRadius(houses, heaters))

File: Heaters3.cpp
#include<iostream>
#include<vector>
#include<unordered_set>
#include<algorithm>

using namespace std;

class Solution {
public:
    int findRadius(vector<int>& houses, vector<int>& heaters) {
    	
    	sort(houses.begin(), houses.end());
		sort(heaters.begin(), heaters.end());
		
		int result = 0;
		int j = 0;
		for(int i = 0; i < houses.size(); i++) {
			int cur = houses[i];
			while(j < heaters.size() - 1 && abs(heaters[j + 1] - cur) <= abs(heaters[j] - cur)) j++;
			result = max(result, abs(heaters[j] - cur));
		}
		
		return result;
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> houses;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		houses.push_back(num);
	}
	
	cin>>n;
	vector<int> heaters;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		heaters.push_back(num);
	}
	
	Solution *solution = new Solution();
	cout<<solution->findRadius(houses, heaters);
	
	return 0;
	
}
